File: Striver/Trees/Q13.java
package Striver.Trees;

public class Q13{
    // max path sum from any node to any node

    public int maxPathSum(Node root) {
        int[] max = new int[1];
        max[0] = Integer.MIN_VALUE;
        maxPathSum(root, max);
        return max[0];
    }

    public int maxPathSum(Node root, int[] max){
        if(root == null) return 0;
        int left = maxPathSum(root.left, max);
        int right = maxPathSum(root.right, max);
        int max_single = Math.max(Math.max(left, right) + root.data, root.data);
        int max_top = Math.max(max_single, left + right + root.data);
        max[0] = Math.max(max[0], max_top);
        return max_single;
    }

    public static void main(String[] args) {
        Q13 sol = new Q13();
        Node root = new Node(10);
        root.left = new Node(2);
        root.right = new Node(10);
        root.left.left = new Node(20);
        root.left.right = new Node(1);
        root.right.right = new Node(-25);
        root.right.right.left = new Node(3);
        root.right.right.right = new Node(4);
        System.out.println(sol.maxPathSum(root));   
    }

}

File: Striver/Trees/Q13.py
# max path sum from any node to any node

from Node import Node

def maxPathSum(root, res):
    if not root:
        return 0
    l = maxPathSum(root.left, res)
    r = maxPathSum(root.right, res)
    temp = max(max(l, r) + root.data, root.data)
    ans = max(temp, l + r + root.data)
    res[0] = max(res[0], ans)
    return temp

def maxPathSumUtil(root):
    if not root:
        return 0
    res = [float('-inf')]
    maxPathSum(root, res)
    return res[0]

root = Node(10)
root.left = Node(2)
root.right = Node(10)
root.left.left = Node(20)
root.left.right = Node(1)
root.right.right = Node(-25)
root.right.right.left = Node(3)
root.right.right.right = Node(4)

print(maxPathSumUtil(root))